Release 4.0.0 alpha.2 (#6479)
open-formulieren/open-forms: CHANGELOG.rst

        `latest <https://open-forms.readthedocs.io/en/latest/changelog.html>`_ docs
        version.

4.0.0-alpha.2 (2026-07-23)
==========================

This is an alpha release, meaning it is not finished yet or suitable for production use.

Detailed changes
----------------

**Breaking changes**

* [:backend:`6164`] Removed deprecated code (please review the instructions in the documentation
  under **Installation** > **Upgrade details to Open Forms 4.0.0** for all the necessary details):

  - [:backend:`4939`] Removed legacy ``zaaktype`` URL and ``documenttype`` URL catalogi

>     configuration options for the Objects API and ZGW APIs registrations.

Reviewer: ```suggestion
  - [:backend:`4939`] Removed legacy ``zaaktype`` and ``documenttype`` URL-based
    configuration options for the Objects API and ZGW APIs registration plugins.
```

>   - [:backend:`5769`] Dropped support for third party extensions.

Reviewer: ```suggestion
  - [:backend:`5769`] The plugin interfaces for extensions are now private API and no longer subject
    to our semantic versioning policy.
```

**New features**

* [:backend:`5764`] Single step forms:

  - [:backend:`6350`] Added option to hide the form title.

>   - [:backend:`6398`] Made url information accessible through a ``form_url`` static variable.

Reviewer: ```suggestion
  - [:backend:`6398`] Added the ``form_url`` static variable, containing metadata about where
    a form submission was started.
```

* [:backend:`6281`] Implemented import converters for file registration options.

> * [:backend:`6399`] Single step forms can now change variable values and evaluate DMN using logic

Reviewer: this should be higher up under the "single step forms" entries

  actions.

> * [:backend:`6343`] Added support for creating case objects in the ZGW APIs registration.

Reviewer: ```suggestion
* [:backend:`6343`] You can now add 'overige' case objects in the ZGW APIs registration plugin.
```

> * [:backend:`6380`] Implemented new ``substr`` JSON logic operator.

Reviewer: ```suggestion
* [:backend:`6380`] The ``substr`` JSON logic operator is now available in logic rules.
```

> * [:backend:`5377`] Updated available Haal Centraal BRP attributes.

Reviewer: ```suggestion
* [:backend:`5377`] Exposed a lot of additional Haal Centraal BRP Personen bevragen prefill attributes.
```

>   - [:backend:`6281`] Implemented import converters for file registration options.

Reviewer: This entry is not part of the single step form :)

**Bugfixes**

* [:backend:`6254`] Fixed pre-3.5 imports of forms that contain logic rules with
  "trigger from step" options.
* [:backend:`6155`] Fixed missing registry hooks for ``customerProfile`` component.
* [:backend:`6275`] Fixed some clear-on-hide issues:

  - Fixed logic not being executed when data is empty.
  - Fixed infinite loop of check-logic calls.

>   - Fixed crash with hidden selectboxes used as condition trigger.

Reviewer: This is not a bugfix entry, but a breaking change entry

```suggestion
* [:backend:`6275`] Changed the behaviour of "clear-on-hide" in logic evaluation - values cleared because the
  component is hidden are now completely absent from the logic evaluation context. Before, they would get the
  component default or empty value for logic evaluation.
```

* [:backend:`6113`, :backend:`5730`] Fixed broken table markup on introduction page
  and/or start page.
* [Sentry#482931] Fixed type error when session data contains ``None``.
* [:backend:`6447`] Fixed visibility calculation during formio validation in ``editgrid``
  components.
* [:backend:`6459`] Fixed templating usage in ``radio``, ``select`` and ``selectboxes`` components.

**Project maintenance**

> * Introduced ``DISABLE_2FA`` environment variable.

Reviewer: ```suggestion
* Re-introduced the ``DISABLE_2FA`` environment variable.
```
can also go into bugfixes section, it was "backported" :)

Author: Ah i look around and couldn't find evidence of it having existed before..

Changing it 👍

> * Updated upgrade check to require version 3.5.4 before updrading to 4.0.0.

Reviewer: ```suggestion
* Updated upgrade check to require version 3.5.4 before upgrading to 4.0.0.
```

Author: 🙈

> * [:backend:`6126`] Hardening CI workflows to protect against supply chain attacks.

Reviewer: ```suggestion
* [:backend:`6126`] Hardened CI workflows to protect against supply chain attacks.
```

* Updated 4.0.0 upgrade notes and form documentation.
* Added Open Forms branding to the admin interface.

> * Updated privileges of admin users.

Reviewer: Can go into features

```suggestion
* Added protections to prevent admin users from modifying their own permissions when they can
  manage (other) admin users.
```

> * [:backend:`6277`] Updated JSON logic to correctly check for missing properties.

Reviewer: You can omit this, it's part of 6121/6275

* Updated backend dependencies to their latest versions:

  - django
  - lxml
  - lxml-html-clean
  - maykin-json-logic-py
  - pydantic-settings
  - pytz
  - soupsieve
  - vcrpy
  - weasyprint
  - zeep

* Updated frontend dependencies:

  - Bumped @open-formulieren/design-tokens to 0.68.1.

* Cleaned up squashed migrations for Yivi attributes, worldline and submission.
* Disabled OAS CI action due to security concerns.
* [:backend:`2713`] Clean up submission file attachment handling and formio configuration path
  usage.
* Replaced ``microscope-sass/lib/bem`` with ``@bbt/bem``.

> * Setting a maximum number of allowed celery worker tasks per child.

Reviewer: ```suggestion
* Added a default for the maximum amount of tasks that may be executed by a celery worker child.
```
